docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -19,161 +20,177 @@ the light client detector receives two conflicting headers
-to the witness and the witness's trace to the primary) as the light client is 
-incapable of deciding which one is malicious. 
-
-Assuming one of the two providers is honest, that full node will then take the trace
-and extract out the relevant evidence and propogate it until it ends up on chain.
-
-*NOTE: We do not send evidence to other connected peers*
-
-*NOTE: The light client halts then and does not verify with any other witness*
-
-## Detailed Design
-
-The traces will have the following data structure: 
+A previously discussed approach to handling evidence was to pass all the data that the 
+light client had witnessed when it had observed diverging headers for the full node to 
+process.This was known as a light trace and had the following structure: 
 
 ```go
 type ConflictingHeadersTrace struct {
   Headers []*types.SignedHeader
 }
 ```
 
-When a full node receives a `ConflictingHeadersTrace`, it should
-a) validate it b) figure out if malicious behaviour is obvious (immediately
-slashable) or run the amnesia protocol.
+This approach has the advantage of not requiring as much processing on the light 
+client side in the event that an attack happens. Although, this is not a significant 
+difference as the light client would in any case have to validate all the headers 
+from both witness and primary. Using traces would consume a large amount of bandwidth
+and adds a ddos vector to the full node.
 
-### Validating headers
 
-Check headers are valid (`ValidateBasic`), are in order of ascending height, and do not exceed the `MaxTraceSize`.
-
-### Finding Block Bifurcation
-
-The node pulls the block ID's for the respective heights of the trace headers from its own block store.
+## Decision
 
-First it checks to see that the first header hash matches its first `BlockID` else it can discard it. 
+The light client will be divided into two components: a `Verifier` (either sequential or 
+bisection) and a `Detector` (see [Informal's Detector](https://github.com/informalsystems/tendermint-rs/blob/master/docs/spec/lightclient/detection/detection.md))
+. The detector will take the trace of headers from the primary and check it against all 
+witnesses. For a witness with a diverging header the, detector will first verify the header
+by bisecting through all the intermediary headers that the primary provided. If valid,
+the light client will trawl through both traces and find the point of bifurcation where it
+can proceed to extract any evidence (as is discussed in detail later). 
 
-If the last header hash matches the nodes last `BlockID` then it can also discard it on the assumption that a fork can not remerge and hence this is just a trace of valid headers. 
+Upon successfully detecting the evidence, the light client will send it to both primary and
+witness before halting. It will not send evidence to other peers nor continue to verify the
+primary's header against any other header.
```

Comment:
Fault detection should ensure that "exposed" faulty header (from a fork) will be detected in a bounded time. This covers the scenario of a faulty primary. However, as Josef pointed out, we should try to cover the scenario of faulty secondary, and submit evidence(s) to a primary (correct) to disincentivize the faulty behaviour. Otherwise, a faulty witness can shut down the light client with a correct primary without any consequences.